atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/AbstractSnapshotTransactionTest.java

```diff
@@ -1989,6 +1989,58 @@ public void canReadWithExpectedSizeOneAfterDeletingPresentCellAndWritingToAnothe
                         any());
     }
 
+    @Test
+    public void throwsRetriableExceptionIfTransactionKeyValueServiceNoLongerValid() {
+        ConjureStartTransactionsResponse startedTransactions = startTransactionWithWatches();
+        long transactionTs = startedTransactions.getTimestamps().start();
+        LongSupplier startTimestampSupplier = () -> transactionTs;
+        InvalidatableTransactionKeyValueService invalidatableTransactionKeyValueService =
+                new InvalidatableTransactionKeyValueService(
+                        txnKeyValueServiceManager.getTransactionKeyValueService(startTimestampSupplier));
+
+        SnapshotTransaction snapshotTransaction = getSnapshotTransactionWith(
+                startTimestampSupplier,
+                startedTransactions.getImmutableTimestamp(),
+                invalidatableTransactionKeyValueService);
+        PathTypeTracker pathTypeTracker = PathTypeTrackers.constructSynchronousTracker();
+        Transaction callbackAwareTransaction = transactionWrapper.apply(snapshotTransaction, pathTypeTracker);
+
+        callbackAwareTransaction.put(TABLE, ImmutableMap.of(TEST_CELL, TEST_VALUE));
+
+        invalidatableTransactionKeyValueService.invalidate();
+        assertThatLoggableExceptionThrownBy(callbackAwareTransaction::commit)
+                .isInstanceOf(TransactionFailedRetriableException.class)
+                .hasLogMessage("Transaction key value service is no longer valid")
+                .args()
+                .contains(SafeArg.of("startTimestamp", transactionTs))
+                .anyMatch(arg -> arg.getName().equals("commitTimestamp") && ((long) arg.getValue()) > transactionTs);
+    }
+
+    @Test
+    public void doesNotThrowExceptionIfTransactionKeyValueServiceNoLongerValidOnTransactionNotPerformingWrites() {
```

Comment:
I think this is correct as per the RFC: nothing about the migration should invalidate a snapshot and WRITE transactions will be forced to abort, MEANING even if we had this check for these transactions and aborted them, this doesn't actually add much, because the write transactions themselves that could've affected the read transactions, will abort. so the abort is pointless and just a perf hit.

Reply:
Yep. I’m assuming that if we do tkvsManager.getTransactionKvs(timestamp), then isValid(timestamp) on the same timestamp should always return true, and reads happen just at a single point in time, so...

Reply:
I guess I do explain it in the assertion: `all their reads effectively at the start timestamp, so these would still be valid.`

Reply:
The only thing that could maybe a problem here is that once we start deleting the source tables, we need to make sure this fails in a very loud and obvious way